rskj-core/src/main/java/org/ethereum/datasource/RocksDbDataSource.java

```diff
@@ -0,0 +1,366 @@
+        if (Paths.get(databaseDir).isAbsolute()) {
+            return Paths.get(databaseDir, name);
+        } else {
+            return Paths.get(getProperty("user.dir"), databaseDir, name);
+        }
+    }
+
+    @Override
+    public boolean isAlive() {
+        try {
+            resetDbLock.readLock().lock();
+            return alive;
+        } finally {
+            resetDbLock.readLock().unlock();
+        }
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public byte[] get(byte[] key) {
+        Objects.requireNonNull(key);
+
+        byte[] result = null;
+
+        Metric metric = profiler.start(Profiler.PROFILING_TYPE.DB_READ);
+        resetDbLock.readLock().lock();
+
+        int retries = 0;
+        Exception exCaught = null;
+
+        try {
+            while (retries < MAX_RETRIES) {
+                try {
+                    if (logger.isTraceEnabled()) {
+                        logger.trace("~> RocksDbDataSource.get(): {}, key: {}", name, ByteUtil.toHexString(key));
+                    }
+
+                    byte[] ret = db.get(key);
+
+                    if (logger.isTraceEnabled()) {
+                        logger.trace("<~ RocksDbDataSource.get(): {}, key: {}, return length: {}", name, ByteUtil.toHexString(key), (ret == null ? "null" : ret.length));
+                    }
+
+                    result = ret;
+                    break;
+                } catch (RocksDBException e) {
+                    logger.error("Exception. Retrying again...", e);
+                    exCaught = e;
+                }
+
+                retries++;
+            }
+        } finally {
+            resetDbLock.readLock().unlock();
+            profiler.stop(metric);
+        }
+
+        if (exCaught != null && retries > 1) {
+            logger.error("Exception. Not retrying.", exCaught);
+            panicProcessor.panic("rocksdb", String.format("Exception. Not retrying. %s", exCaught.getMessage()));
+            throw new RuntimeException("Couldn't get the data back for the given key");
+        }
+
+        return result;
+    }
+
+    @Override
+    public byte[] put(byte[] key, byte[] value) {
+        Objects.requireNonNull(key);
+        Objects.requireNonNull(value);
+
+        Metric metric = profiler.start(Profiler.PROFILING_TYPE.DB_WRITE);
+        resetDbLock.readLock().lock();
+
+        try {
+            if (logger.isTraceEnabled()) {
+                logger.trace("~> RocksDbDataSource.put(): {}, key: {}, return length: {}", name, ByteUtil.toHexString(key), value.length);
+            }
+
+            db.put(key, value);
+
+            if (logger.isTraceEnabled()) {
+                logger.trace("<~ RocksDbDataSource.put(): {}, key: {}, return length: {}", name, ByteUtil.toHexString(key), value.length);
+            }
+        } catch (RocksDBException e) {
+            logger.error("Exception. Not retrying.", e);
```

Comment:
probably we should propagate the exception up here?

Reply:
Done